File: test/ol/trixnity_poc/bot_logic_test.clj
(ns ol.trixnity-poc.bot-logic-test
  (:require
   [clojure.test :refer [deftest is testing]]
   [ol.trixnity-poc.bot-logic :as sut])
  (:import
   (net.folivo.trixnity.core.model EventId UserId)
   (net.folivo.trixnity.core.model.events.m ReactionEventContent RelatesTo$Annotation)))

(deftest mirrored-body-test
  (is (= "HELLO MATRIX"
         (sut/mirrored-body "Hello Matrix"))))

(deftest should-handle-sender-test
  (let [bot-user (UserId. "@bot:example.org")]
    (is (false? (sut/should-handle-sender? bot-user bot-user)))
    (is (true? (sut/should-handle-sender?
                (UserId. "@alice:example.org")
                bot-user)))))

(deftest reaction-to-mirror-test
  (testing "returns target event id and emoji key"
    (let [content  (ReactionEventContent.
                    (RelatesTo$Annotation.
                     (EventId. "$event:example.org")
                     "👍")
                    nil)
          mirrored (sut/reaction-to-mirror content)]
      (is (some? mirrored))
      (is (= (EventId. "$event:example.org")
             (:event-id mirrored)))
      (is (= "👍"
             (:key mirrored)))))

  (testing "returns nil when reaction has no annotation relation"
    (is (nil? (sut/reaction-to-mirror (ReactionEventContent.))))))

File: kotlin/src/test/kotlin/phase1/BotLogicTest.kt
package phase1

import net.folivo.trixnity.core.model.EventId
import net.folivo.trixnity.core.model.UserId
import net.folivo.trixnity.core.model.events.m.ReactionEventContent
import net.folivo.trixnity.core.model.events.m.RelatesTo
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class BotLogicTest {
    @Test
    fun `mirroredBody uppercases text`() {
        assertEquals("HELLO MATRIX", BotLogic.mirroredBody("Hello Matrix"))
    }

    @Test
    fun `shouldHandleSender ignores bot's own events`() {
        val botUser = UserId("@bot:example.org")
        assertFalse(BotLogic.shouldHandleSender(botUser, botUser))
        assertTrue(BotLogic.shouldHandleSender(UserId("@alice:example.org"), botUser))
    }

    @Test
    fun `reactionToMirror returns target event id and emoji key`() {
        val content = ReactionEventContent(
            relatesTo = RelatesTo.Annotation(
                eventId = EventId("\$event:example.org"),
                key = "👍",
            ),
        )

        val mirrored = BotLogic.reactionToMirror(content)
        assertNotNull(mirrored)
        assertEquals(EventId("\$event:example.org"), mirrored.eventId)
        assertEquals("👍", mirrored.key)
    }

    @Test
    fun `reactionToMirror returns null when reaction has no annotation relation`() {
        val content = ReactionEventContent(relatesTo = null)
        assertNull(BotLogic.reactionToMirror(content))
    }
}
